PR1 - Hydrogen-react

juanpprieto opened the pull request:
### WHY are these changes introduced?

During the React Router 7.8.x migration, we encountered TypeScript errors when passing values between Storefront API and Customer Account API due to enum type incompatibilities. The initial approach forced Customer Account API to use Storefront API's enum types, which prevented the Customer Account API from supporting its own specific values like the USDC currency code.

This caused issues because:
- Customer Account API can return currency codes (like USDC) that aren't in Storefront API's CurrencyCode enum
- The forced type sharing prevented proper runtime handling of these API-specific values
- Tests that validated USDC handling were failing with TypeScript errors

### WHAT is this pull request doing?

This PR restores proper enum type handling for Customer Account API by:

1. **Removing the `enumValues` override** in `codegen.ts` that forced Customer Account API to use Storefront API types
2. **Allowing Customer Account API to generate its own enum types**, including API-specific values like USDC currency
3. **Leveraging existing union type support** in Money component and useMoney hook to handle both API types properly

The Money component and useMoney hook already use union types (`StorefrontApiMoneyV2 | CustomerAccountApiMoneyV2`) to accept data from either API, and the hook includes runtime detection for unsupported currency codes with graceful fallback formatting.

**Changes:**
- `packages/hydrogen-react/codegen.ts`: Removed enumValues override, allowing separate enum generation
- `packages/hydrogen-react/src/Money.test.helpers.ts`: Updated to use USDC as example of Customer Account API-specific currency
- `.changeset/fix-enum-compatibility.md`: Added changeset for this fix

### HOW to test your changes?

1. **Verify TypeScript compilation:**
   ```bash
   cd packages/hydrogen-react
   npm run typecheck
   ```

2. **Run tests to ensure USDC handling works:**
   ```bash
   cd packages/hydrogen-react
   npm test
   ```

3. **Test in a Hydrogen app:**
   - Use Customer Account API mutations with `i18n.language` from Storefront context
   - Render Money components with Customer Account API data that includes USDC currency
   - Verify no TypeScript errors and proper runtime formatting

4. **Verify enum generation:**
   ```bash
   cd packages/hydrogen-react
   npm run graphql-types
   # Check that src/customer-account-api-types.d.ts includes USDC in CurrencyCode
   grep "| 'USDC'" src/customer-account-api-types.d.ts
   ```

#### Post-merge steps

None required. Type generation is automatic during the build process.

#### Checklist

- [x] I've read the [Contributing Guidelines](https://github.com/Shopify/hydrogen/blob/main/CONTRIBUTING.md)
- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've added a [changeset](https://github.com/Shopify/hydrogen/blob/main/CONTRIBUTING.md#changesets) if this PR contains user-facing or noteworthy changes
- [x] I've added [tests](https://github.com/Shopify/hydrogen/blob/main/CONTRIBUTING.md#testing) to cover my changes
- [x] I've added or updated the documentation (inline comments explain the approach)

juanpprieto committed:
[2025-07] Fix Money component compatibility with Customer Account API USDC currency (#3090)

* Fix Money component compatibility with Customer Account API USDC currency

The 2025-07 API update introduced USDC currency code to Customer Account API but not Storefront API, causing TypeScript errors and runtime failures.

Changes:
- Update Money component to accept MoneyV2 from both Storefront and Customer Account APIs via union types
- Enhance useMoney hook to detect unsupported currency codes (like USDC) and gracefully fall back to decimal formatting
- Add currency code suffix for unsupported currencies (e.g., "100.00 USDC") to maintain clarity
- Default to 2 decimal places for USDC to reinforce its 1:1 USD peg based on industry standards

Technical details:
- Handle Intl.NumberFormat RangeError for cryptocurrency codes not in ISO 4217
- Add comprehensive test coverage for both API types and USDC formatting
- Update TypeScript types to support both CurrencyCode enums

Fixes #3089

* Add changeset for Money component USDC compatibility fix

* Add build script to copy customer-account-api-types to dist

Since our Money component now imports from customer-account-api-types, we need to ensure this file is copied to the dist folder during the build process. Previously, only storefront-api-types was being copied.

- Replace copy-storefront-types with copy-api-types script that copies both API type files
- Update build and dev:demo scripts to use the new copy-api-types script
- This fixes the CI build failure where customer-account-api-types.d.ts was not found in dist

* Trigger CI rebuild

kdaviduik committed:
Suggested money changes for PR #3090 (#3097)

juanpprieto committed:
feat: pin React Router to exact version 7.8.2 across monorepo

- Pin react-router and react-router-dom to exact 7.8.2
- Pin @react-router/* packages to exact 7.8.2
- Remove version ranges (~, ^) to ensure consistency
- Update all packages, templates, and examples
- Update package-lock.json to lock exact versions

juanpprieto committed:
fix: hoist @react-router/dev and fs-routes to fix CLI tests

- Add @react-router/dev and @react-router/fs-routes to root devDependencies
- Ensures packages are available in root node_modules for test symlinks
- Fixes CLI setup test that creates temporary directories

juanpprieto committed:
fix: update .gitignore and example lint scripts for React Router 7.8.2 compatibility

- Add .react-router to .gitignore (generated by React Router 7.8.2)
- Remove deprecated --ext flag from example lint scripts for ESLint 9 compatibility

juanpprieto committed:
fix: update UIMatch test helper for React Router 7.8.2 compatibility

- React Router 7.8.2 requires loaderData to be non-optional in UIMatch type
- Add loaderData property to fillMatch helper in seo.test.ts

juanpprieto committed:
Lint and format

juanpprieto committed:
Lint and format

juanpprieto committed:
Add changeset

juanpprieto committed:
fix: TypeScript 5.9 compatibility updates

- Upgrade TypeScript to 5.9.2 across all packages
- Fix enum compatibility between Storefront and Customer Account APIs
- Fix Navigation type mocks in hydrogen package tests
- Fix BodyInit type compatibility in cache implementation

juanpprieto committed:
fix: resolve eslint dependency issue in subscriptions example

Add missing dependencies to useMemo hook to satisfy react-hooks/exhaustive-deps rule

juanpprieto committed:
fix: properly fix useMemo dependencies in subscriptions example

Move params initialization inside useMemo to avoid recreating on every render

juanpprieto committed:
fix: resolve TypeScript 5.9 compatibility issues with ESLint and update dependencies

This commit addresses critical TypeScript 5.9 compatibility issues that were blocking
the React Router 7.8.x migration. The changes ensure all CI checks pass with zero errors.

## Key Changes:

### ESLint TypeScript Compatibility
- Removed unused `eslint-plugin-hydrogen` package that was causing TypeScript version conflicts
- Updated `@typescript-eslint/eslint-plugin` and `@typescript-eslint/parser` to v8.42.0
- Pinned all ESLint-related package versions for consistency and reproducibility

### Fixed Lint Errors Across Examples
- GTM example: Added missing `ready` and `subscribe` dependencies to useEffect hooks
- Partytown example: Added missing `location.pathname` dependency, fixed iframe accessibility
- Subscriptions example: Replaced deprecated @ts-ignore with @ts-expect-error directives
- Added missing iframe title attributes for accessibility compliance

### Build Configuration
- Added `dist/` to eslint ignores in skeleton template to prevent linting of built files
- Standardized ESLint configuration across all examples using flat config format
- Created eslint.config.js files for all examples to ensure consistent linting

### Dependencies
- Added `@total-typescript/ts-reset@0.6.1` at monorepo level (pinned version)
- Fixed duplicate import issue in hydrogen/src/seo/seo.ts

## Technical Details:

The root cause was eslint-plugin-hydrogen depending on an older version of
@typescript-eslint packages that don't support TypeScript 5.9. Since the plugin
wasn't actively used in the codebase, removing it was the cleanest solution.

All ESLint packages are now pinned to specific versions to prevent future
compatibility issues during dependency updates.

## Testing:
- ✅ All lint checks pass
- ✅ All typecheck passes
- ✅ No ESLint warnings about TypeScript version compatibility
- ✅ All examples build successfully

juanpprieto committed:
ci: trigger CI workflows

juanpprieto committed:
chore: remove pr-plans and changeset files from migration branch

These files were added during development but should not be part of the final PR:
- Removed .github/pr-plans/ directory (planning documentation)
- Removed .changeset/ files added during development

These are internal development artifacts not needed for the React Router 7.8.x migration.

juanpprieto committed:
fix(hydrogen-react): update codegen.ts to use Storefront API types for Customer Account enums

- Configure Customer Account API codegen to reference Storefront API types for LanguageCode and CurrencyCode enums
- Ensures type compatibility between APIs when passing i18n.language from Storefront to Customer Account API
- Regenerate Customer Account API types with proper enum references
- Resolves TypeScript errors in skeleton template account routes

juanpprieto committed:
fix: address PR review feedback

- Remove CI_ISSUES_REPORT.md and PR1_STATUS.md files
- Restore deleted changeset file dry-swans-relate.md
- Improve comment for SellingPlanGroup type assertion

juanpprieto commented on packages/hydrogen-react/src/storefront-api-types.d.ts:
```diff
@@ -708,7 +708,7 @@ export type CartBuyerIdentity = {
    * The rank of the preferences is determined by the order of the addresses in the array. Preferences
    * can be used to populate relevant fields in the checkout flow.
    *
-   * As of the `2025-04` release, `buyerIdentity.deliveryAddressPreferences` is deprecated.
```
Is this correct? Can we validate when this was actually deprecated ?

juanpprieto committed:
fix: address PR review feedback

- Remove .claude/commands/shopify-cli-update.md file
- Regenerate GraphQL types to ensure correctness
- Add changeset for enum compatibility fix

juanpprieto committed:
chore: merge feat/typescript-lint-rr-7.8 into feat/hydrogen-react-rr-7.8

- Resolved conflicts in codegen.ts by merging both approaches
- Used typescript-lint version of customer-account-api-types.d.ts
- Added CountryCode to enumValues configuration
- Restored .claude/commands/shopify-cli-update.md file

kdaviduik commented on packages/hydrogen-react/codegen.ts:
```diff
@@ -88,8 +88,13 @@ const config: CodegenConfig = {
+            // Use Storefront API's enum types for Customer Account API to ensure type compatibility.
+            // Both APIs use the same 2025-04 schema version and share 95%+ of enum values.
+            // The Storefront API may have additional codes that Customer Account API doesn't define,
+            // but using the superset is safe - Customer Account will accept all values it supports.
```
just to confirm, this means that the Customer Account API will accept all values that the SF API defines, even if those values aren't actually defined in the CA API?

rbshop commented:
Looks like you've changed what this PR actually does but not updated the description or title. Please can you update those

juanpprieto committed:
fix(hydrogen-react): restore separate enum types for Customer Account API

- Remove enumValues override that forced Customer Account API to use Storefront API types
- Allow Customer Account API to generate its own enum types (including USDC currency)
- Update Money component and useMoney hook to handle union types properly
- Fix test helper to use USDC as example of Customer Account API-specific currency
- This properly supports runtime values like USDC that Customer Account API can return

juanpprieto committed:
chore: remove unrelated shopify-cli-update.md changes

This file was inadvertently modified in commit 054c2b39d and should match the base branch version

kdaviduik commented on packages/hydrogen-react/codegen.ts:
```diff
@@ -88,13 +88,8 @@ const config: CodegenConfig = {
-              CurrencyCode: '#CurrencyCode',
-              CountryCode: '#CountryCode',
-            },
+            // Generate separate enum types for Customer Account API
```
Earlier in this file the API version is set to 2025-04 for both APIs, shouldn't that be 2025-07 now?

<img width="582" height="286" alt="Image" src="https://github.com/user-attachments/assets/8fb1edd6-ac34-4394-bec2-b1ae4ddcb4bf" />

juanpprieto replied:
not on this branch

kdaviduik replied:
ah yes!

juanpprieto committed:
Merge branch 'main' into feat/hydrogen-react-rr-7.8

Resolved conflicts:
- package.json: Included react/react-dom dependencies from main
- packages/cli/package.json: Added @remix-run/dev from main
- packages/hydrogen-react/codegen.ts: Kept branch version (core USDC support)
- packages/hydrogen-react/src/Money.test.helpers.ts: Kept descriptive USDC comment
- packages/hydrogen-react/src/Money.test.tsx: Took main's version with unit price tests
- templates/skeleton/package.json: Updated to CLI ~3.83.3 from main
- Schema files: Accepted main's 2025-07 API versions
- package-lock.json: Regenerated fresh
